### ISSUE TITLE
Documentation on Slice, LinearIndices, etc. with custom array indexing

### ISSUE THREAD

**Issue by damiendr** [JuliaLang/julia#36830]
It would be nice to have some documentation on the use of `Slice`, `LinearIndices` and `CartesianIndices` for non-1-based indexing.

For instance I just found out in #28503 that the following works:
```julia
julia> li = LinearIndices((Base.Slice(1:128), Base.Slice(0:127)))
julia> li[1,0]
1
```

Until then I had unsuccessfully tried the following:
```julia
julia> li = LinearIndices((1:128, 0:127))
julia> li = LinearIndices((1:128, ZeroRange(128)))
```
which both give 
```
julia> li[1,0]
ERROR: BoundsError: attempt to access 128×128 LinearIndices{2,Tuple{UnitRange{Int64},UnitRange{Int64}}} at index [1, 0]
```

But I can't find any official docs on `Slice` or why it works differently from `ZeroRange` and `UnitRange`, or an explanation of the rationale for the implicit 1-based conversions.

Example of things I would have liked to find in the docs:

- what design choices require that `LinearIndices((0:9,0:9)) == LinearIndices((1:10, 1:10))`?

- if I can do:
```julia
julia> CartesianIndices((Base.Slice(0:9), Base.Slice(0:9)))[0,0]
CartesianIndex(0, 0)
julia> CartesianIndices((0:9, 0:9))[1,1]
CartesianIndex(0, 0)
```
is there a built-in way to construct an expression `SomeIndices` such that:
```julia
julia> SomeIndices((0:9, 0:9))[0,0]
CartesianIndex(1,1)
```
other than allocating a complete `OffsetArray` with the indices?

Maybe this belongs to the dev/offset arrays docs?

**Comment by mbauman** [comment]
Yes, this can and should be improved.  Quick comment right off the bat: you want to use `Base.IdentityUnitRange` instead of `Base.Slice`.  I'll try to add more later, but some details are in discourse:

* https://discourse.julialang.org/t/stubborn-1-based-linearindices-for-the-custom-type-with-non-zero-based-indexing/40593/6

* https://discourse.julialang.org/t/linearindices-for-non-1-based-indices/26906/2
